### ISSUE TITLE
How to get the feature bounded by the detected box in object detection?

### ISSUE THREAD

**Issue by Jian-Xi** [ultralytics/ultralytics#6311]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

Hi! I'm working on an object detection task.

To train my fusion model, I'd like to extract the features bounded by the detected boxes, not the boxes themselves, rather than the pixel values of detected objects like dogs, cats, etc, i.e., the contents of detected bounding boxes. 

I wonder if this is possible. I did not find anything about this in the instance object (Result). Probably it should be done before postprocessing. I'm checking this thread [5192](https://github.com/ultralytics/ultralytics/issues/5192), but it's somehow different from what I intended.

In short word, how to get the feature bounded by the detected box?
Thanks in advance!

**Comment by RizwanMunawar** [comment]
@Jian-Xi you can use the mentioned code to extract the pexels data of bounding boxes (pixel values).

```python
import cv2
from ultralytics import YOLO

model = YOLO("..\\weights\\yolov8s.pt")
results = model("https://ultralytics.com/images/bus.jpg")

boxes = results[0].boxes.xywh.cpu()
im0 = cv2.imread("bus.jpg")

for [x1, y1, x2, y2] in boxes:
    object_cropped = im0[int(y1):int(y2), int(x1):int(x2)]
    print(object_cropped)

```

Hope this helps! Let me know if you have any additional questions.
Thanks

**Comment by Jian-Xi** [comment]
@RizwanMunawar Thank you very much! This is of course a possible solution. What I'm thinking is that some operations can be done at least (before) the 23rd layer, because in some cases the model may not detect objects after NMS. This doesn't mean that there are no objects. For example, for image captioning, the visual features are taken usually before NMS (see [Paper](https://github.com/peteanderson80/bottom-up-attention)). So I want some sort of intermediate feature map before the model makes the final decisions. 

Thanks

**Comment by glenn-jocher** [comment]
Understood @Jian-Xi, you're looking to tap into intermediate model layers to obtain feature maps before the final Non-Maximum Suppression (NMS) is applied. This can indeed be useful for tasks like image captioning where you're interested in the raw features extracted by the convolutional layers.

In YOLOv8, you can accomplish this by intercepting the output of the desired layer before the post-processing steps. The design of YOLOv8 allows for access to internal layers through model attributes or by modifying the forward method. What you would do is run a forward pass up to your layer of interest, which will give you the feature map that corresponds to the raw detections. You can then use these feature maps however you see fit in your fusion model or for other purposes.

To figure out which layer to access, you might need to examine the model architecture and determine the correct layer index or name that contains the type and depth of features you're interested in.

Remember, since this involves modifying the model's internals, you should have a solid understanding of PyTorch and the YOLOv8 architecture to effectively implement these changes. If you require further conceptual clarification, you can refer to the feature extraction section in the Ultralytics documentation or seek guidance from the community on discussions and issues within the YOLOv8 repository.

**Comment by Jian-Xi** [comment]
Since the detection heads accept the intermediate features mapped from the 15/18/21st layers, the bounding box regressor takes the output of the convolution layers (e.g., cv2, cv3) in headblocks for predicting bounding boxes. The output of cv2 and cv3 could be helpful in my case. Here is the possible solution (just modified the **non_max_suppression** in the file **ultralytics/utils/ops.py** ) for extracting the intermediate features in the bounding box. I hope the logic here is more or less correct (@glenn-jocher, any input here is appreciated) and beneficial. Modification is marked with _Detected bbox features_.
`
def non_max_suppression(
        prediction,
        conf_thres=0.25,
        iou_thres=0.45,
        classes=None,
        agnostic=False,
        multi_label=False,
        labels=(),
        max_det=300,
        nc=0,  # number of classes (optional)
        max_time_img=0.05,
        max_nms=30000,
        max_wh=7680,
):
    """
    Perform non-maximum suppression (NMS) on a set of boxes, with support for masks and multiple labels per box.
    Returns:
        (List[torch.Tensor]): A list of length batch_size, where each element is a tensor of
            shape (num_boxes, 6 + num_masks) containing the kept boxes, with columns
            (x1, y1, x2, y2, confidence, class, mask1, mask2, ...).
    """

    # Checks
    assert 0 <= conf_thres <= 1, f'Invalid Confidence threshold {conf_thres}, valid values are between 0.0 and 1.0'
    assert 0 <= iou_thres <= 1, f'Invalid IoU {iou_thres}, valid values are between 0.0 and 1.0'
    if isinstance(prediction, (list, tuple)):  # YOLOv8 model in validation model, output = (inference_out, loss_out)
        feats = prediction[1] # Detected bbox features Add
        prediction = prediction[0]  # select only inference output

    device = prediction.device
    mps = 'mps' in device.type  # Apple MPS
    if mps:  # MPS not fully supported yet, convert tensors to CPU before NMS
        prediction = prediction.cpu()
    bs = prediction.shape[0]  # batch size
    nc = nc or (prediction.shape[1] - 4)  # number of classes
    nm = prediction.shape[1] - nc - 4
    mi = 4 + nc  # mask start index
    xc = prediction[:, 4:mi].amax(1) > conf_thres  # candidates

    # Settings
    # min_wh = 2  # (pixels) minimum box width and height
    time_limit = 0.5 + max_time_img * bs  # seconds to quit after 0.5
    multi_label &= nc > 1  # multiple labels per box (adds 0.5ms/img)

    prediction = prediction.transpose(-1, -2)  # shape(1,84,6300) to shape(1,6300,84)
    prediction[..., :4] = xywh2xyxy(prediction[..., :4])  # xywh to xyxy

    t = time.time()
    output = [torch.zeros((0, 6 + nm), device=prediction.device)] * bs
    output_bbox_feats = list() # Detected bbox features holder
    for xi, x in enumerate(prediction):  # image index, image inference
        # Apply constraints
        # x[((x[:, 2:4] < min_wh) | (x[:, 2:4] > max_wh)).any(1), 4] = 0  # width-height
        x = x[xc[xi]]  # confidence

        # Cat apriori labels if autolabelling
        if labels and len(labels[xi]):
            lb = labels[xi]
            v = torch.zeros((len(lb), nc + nm + 4), device=x.device)
            v[:, :4] = xywh2xyxy(lb[:, 1:5])  # box
            v[range(len(lb)), lb[:, 0].long() + 4] = 1.0  # cls
            x = torch.cat((x, v), 0)

        # If none remain process next image
        if not x.shape[0]:
            continue

        # Detections matrix nx6 (xyxy, conf, cls)
        box, cls, mask = x.split((4, nc, nm), 1)

        # Detected bbox features Add-S
        xfeat = [feats[i][xi,...].unsqueeze(dim=0) for i in range(len(feats))]  # Detected bbox there detection heads
        x_cat = torch.cat([xi.view(xfeat[0].shape[0], nc + 16 * 4, -1) for xi in xfeat], 2)
        box_feats, cls_feats = x_cat.split((16 * 4, nc), 1)
        box_feats = box_feats.transpose(-1, -2).squeeze(0)
        box_feats = box_feats[xc[xi]]
       # Detected bbox features Add-E

        if multi_label:
            i, j = torch.where(cls > conf_thres)
            x = torch.cat((box[i], x[i, 4 + j, None], j[:, None].float(), mask[i]), 1)
        else:  # best class only
            conf, j = cls.max(1, keepdim=True)
            x = torch.cat((box, conf, j.float(), mask), 1)[conf.view(-1) > conf_thres]

        # Filter by class
        if classes is not None:
            x = x[(x[:, 5:6] == torch.tensor(classes, device=x.device)).any(1)]

        # Check shape
        n = x.shape[0]  # number of boxes
        if not n:  # no boxes
            continue
        if n > max_nms:  # excess boxes
            x = x[x[:, 4].argsort(descending=True)[:max_nms]]  # sort by confidence and remove excess boxes

        # Batched NMS
        c = x[:, 5:6] * (0 if agnostic else max_wh)  # classes
        boxes, scores = x[:, :4] + c, x[:, 4]  # boxes (offset by class), scores
        i = torchvision.ops.nms(boxes, scores, iou_thres)  # NMS
        i = i[:max_det]  # limit detections

        # # Experimental
        # merge = False  # use merge-NMS
        # if merge and (1 < n < 3E3):  # Merge NMS (boxes merged using weighted mean)
        #     # Update boxes as boxes(i,4) = weights(i,n) * boxes(n,4)
        #     from .metrics import box_iou
        #     iou = box_iou(boxes[i], boxes) > iou_thres  # iou matrix
        #     weights = iou * scores[None]  # box weights
        #     x[i, :4] = torch.mm(weights, x[:, :4]).float() / weights.sum(1, keepdim=True)  # merged boxes
        #     redundant = True  # require redundant detections
        #     if redundant:
        #         i = i[iou.sum(1) > 1]  # require redundancy

        output[xi] = x[i]
        box_feats = box_feats[i]  # pick the bbox feature after nmx confidence
        print(f'box_feats.shape:{box_feats.shape}')
        output_bbox_feats.append(box_feats)
        if mps:
            output[xi] = output[xi].to(device)
        if (time.time() - t) > time_limit:
            LOGGER.warning(f'WARNING ⚠️ NMS time limit {time_limit:.3f}s exceeded')
            break  # time limit exceeded

    return output, output_bbox_feats
`

**Comment by glenn-jocher** [comment]
@Jian-Xi you're correct in pointing out that intermediate feature maps can be very useful for certain tasks, such as the one you mentioned from the image captioning paper. It seems you have delved into the model architecture to identify relevant layers for extracting these features.

Looking at the code snippet you provided, you're attempting to extract the feature maps corresponding to bounding boxes directly from the prediction heads, which indeed hold the pre-NMS feature information. This approach can work, however, be aware that direct modifications to the core functions, such as `non_max_suppression`, can lead to potential maintenance issues when the model updates or changes in structure.

Instead, a more maintainable approach is to create a custom forward method or function, which not only does the usual forward pass but also returns the intermediate features from the desired layers. You could tap into the layers (e.g., cv2, cv3, or any other layers that carry the feature information you are interested in) without modifying the NMS directly.

To preserve the originality of the function, you might consider subclassing the base model to extend its functionalities. This way, you could add methods to retrieve feature maps at the inference stages you're interested in, while still being able to update the original base files from the Ultralytics repository without losing your custom work.

Please make sure you have a proper understanding of the model architecture and be cautious with changes, as they might affect other parts of the post-processing pipeline. If you need more guidance on interpreting and modifying the architecture within the scope of YOLOv8, the Ultralytics documentation along with community discussions can be a great resource to ensure your modifications are sound and stable.

It’s really inspiring to see community members diving into the internals of the YOLO models and coming up with innovative ideas to extend their functionalities.
